### DISCUSSION TITLE
A public supported device database.

### DISCUSSION THREAD

**Discussion by marcus-j-davies** [zwave-js/zwave-js#1359] [Ideas] [answered]
What's your thoughts on having a device support database, that is based on the config package.
one could create a web UI/Interface, that has a searchable database, purely based on the config contents.

hosted on GitHub pages or otherwise.

Each result could list all the params, metadata for the device, and so on.

This could facilitate [#1245](https://github.com/zwave-js/node-zwave-js/issues/1245) allowing end users to discover the proprietary controller functions.

As its based on the config package, it should mirror device support pretty accurately.
It's something I wont mind experimenting with, and to contribute to the project' development

**Comment by AlCalzone** [comment, marked as answer]
This was already proposed here: https://github.com/zwave-js/zwavejs2mqtt/issues/148

If you'd like to give it a shot, go ahead! We should discuss some of the requirements though.

**Comment by marcus-j-davies** [comment]
Yup I think it could be a handy tool for sure. I will review that thread in more detail now, and ask questions/discuss.

**Comment by marcus-j-davies** [comment]
@AlCalzone, @robertsLando 

Have any objections for me to create an org scope private repo for this?
Have been doing some reading on github' oauth provider, and it should work really well to protect the database from spam/random folk.

https://docs.github.com/en/free-pro-team@latest/developers/apps/authorizing-oauth-apps

The idea is that we register the app (in our case the web app),
with this, we get a `client_secret`.

to authenticate users wishing to flag working devices (or anything else we dream up)
we send them to GitHub auth page (where it asks the user permission for our app to verify them)
once granted, it diverts back to our web app, but with an  appended `code` parameter.

This `code `parameter is used to gain an access token, for which we can then obtain the users name and basic info using our `client_secret`
we can then use that info to : not allow updates if not logged in via github + associate a device being flagged to the user.

lots more ground work, R&D of course, but the foundations are there.
we can use Heroku to host the express app/take care of creating the PR's and all the wizardry.

I would like to develop some proof of concepts before getting into it too much.

**Reply by AlCalzone** [reply]
I can register the repo and app (in fact I'm already hosting a private app on Heroku ;) )
Just let me know which information I should enter.

**Reply by marcus-j-davies** [reply]
Our app will be considered an `oauth `app.

to register it : https://github.com/settings/applications/new

but it is asking for URL's such as the callback URL.
so some ground work is likely needed before we can register. - we might need to wait until we get something worth testing, as we will need to understand the URLs we get from Heroku.

having a repo to chuck code at in the time, however, will be the kick start.

**Reply by marcus-j-davies** [reply]
Heroku....

Have only read a little so far,
but does the nodejs apps needs to be any different i.e special package parameters for Heroku to run then?

I assume your using a free/hobby dyno?
does it allow more than 1 app (I'm guessing the resource is shared across apps)

just gathering info at this point.

**Reply by AlCalzone** [reply]
I can create multiple apps and even pipelines, but free apps are limited to personal accounts and single dynos per app.

**Comment by marcus-j-davies** [comment]
Unanswered as I have a mental tick about it whilst in discussion

**Comment by AlCalzone** [comment]
Nothing special is needed, just a package.json file with a `"start"` script that is executed when the dyno boots.
You can also build TypeScript if you have the dependencies listed and a `"build"` script.

The private key (and possibly other options) can be configured in the UI.

**Reply by AlCalzone** [reply]
Let's agree on an app name, then I'll create the project on Heroku and the repo here.
`@zwave-js/config-db-browser`?

**Reply by marcus-j-davies** [reply]
`config-db-browser` is good.
I have created a private repo already, so just change its name over at will.

TypeScript : i am not overly experienced yet, so not familiar enough to work with it yet.
given this is not a library, so type safety isn't much of a concern, do you have any objections?

**Reply by AlCalzone** [reply]
You're building it, so you do the stack decisions :)

**Reply by AlCalzone** [reply]
Could you join our slack, so I don't have to post the backend URLs here? ;)
https://join.slack.com/t/zwave2mqtt/shared_invite/enQtNjc4NjgyNjc3NDI2LTc3OGQzYmJlZDIzZTJhMzUzZWQ3M2Q3NThmMjY5MGY1MTc4NjFiOWZhZWE5YjNmNGE0OWRjZjJiMjliZGQyYmU
